src/controllers/dynakube/activegate/secrets/authtoken_reconciler.go

```diff
@@ -15,52 +16,66 @@ import (
 )
 
 const (
-	ActiveGateAuthTokenName = "auth-token"
+	ActiveGateAuthTokenName   = "auth-token"
+	AuthTokenRotationInterval = time.Hour * 24 * 30
 )
 
 type AuthTokenReconciler struct {
 	client.Client
 	apiReader client.Reader
 	instance  *dynatracev1beta1.DynaKube
 	scheme    *runtime.Scheme
-	apiToken  string
 	dtc       dtclient.Client
 }
 
-func NewAuthTokenReconciler(clt client.Client, apiReader client.Reader, scheme *runtime.Scheme, instance *dynatracev1beta1.DynaKube, apiToken string, dtc dtclient.Client) *AuthTokenReconciler {
+func NewAuthTokenReconciler(clt client.Client, apiReader client.Reader, scheme *runtime.Scheme, instance *dynatracev1beta1.DynaKube, dtc dtclient.Client) *AuthTokenReconciler {
 	return &AuthTokenReconciler{
 		Client:    clt,
 		apiReader: apiReader,
 		scheme:    scheme,
 		instance:  instance,
-		apiToken:  apiToken,
 		dtc:       dtc,
 	}
 }
 
 func (r *AuthTokenReconciler) Reconcile() error {
-	_, err := r.createSecretIfNotExists()
+	_, err := r.reconcileAuthTokenSecret()
 	if err != nil {
 		return errors.Errorf("failed to create activeGateAuthToken secret: %v", err)
 	}
 
 	return nil
 }
 
-func (r *AuthTokenReconciler) createSecretIfNotExists() (*corev1.Secret, error) {
+func (r *AuthTokenReconciler) reconcileAuthTokenSecret() (*corev1.Secret, error) {
 	var config corev1.Secret
 	err := r.apiReader.Get(context.TODO(),
 		client.ObjectKey{Name: r.instance.ActiveGateAuthTokenSecret(), Namespace: r.instance.Namespace},
 		&config)
-	if k8serrors.IsNotFound(err) {
-		log.Info("creating activeGateAuthToken secret")
-		agSecretData, err := r.getActiveGateAuthToken()
-		if err != nil {
-			return nil, errors.Errorf("failed to create secret '%s': %v", extendWithAGSecretSuffix(r.instance.Name), err)
+	if err != nil {
+		if k8serrors.IsNotFound(err) {
+			log.Info("creating activeGateAuthToken secret")
+			return r.ensureAuthTokenSecret()
+		}
+		return nil, errors.WithStack(err)
+	}
+	if isSecretOutdated(&config) {
+		log.Info("activeGateAuthToken is outdated, creating new one")
+		if err := r.deleteSecret(&config); err != nil {
```

Comment:
Instead of deleting the secret, it could be updated instead to save the delete permission

Reply:
yes, I had a discussion with @0sewa0 about that, the downside of updating the secret is that we need to save the timeStamp, when the authToken got updated in order to track, when we need to update the token again

Reply:
makes sense